src/msg/message.cc

```diff
@@ -19,44 +23,31 @@ class SubscriptionInfo {
     std::vector<CRNode*>             sub_list_;
```

Comment:
I think vector should be ok because we don't care the sub/unsub performance that much, but want faster iteration. A simpler data structure fits better imo.

Reply:
You can have both.
Add a `map<const CRNode*, size_t>` for indexing?

Reply:
I don't think we nned to do that, at least not now. Performance of sub/unsub is not on the critical path, so linear time is acceptable, but adding a map for index means I need to add more code to maintain it, probably adding member functions to SubscriptionInfo class for lookup/insert/remove operations. I guess that is a little bit over-engineering.

Reply:
> probably adding member functions to SubscriptionInfo class for lookup/insert/remove operations

Yes that's over engineering, I would probably just call both member vars directly....

Anyway, this one is up to you if perf doesn't matter here.
Also, yes scanning a vector is faster than map, but just in case, map scanning is also linear by std if you didn't realize, 因为你跑了个欧拉刻画

Reply:
Yes, I know map iteration is linear time ...